records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,187 @@
+   slowing down due to a slower database, the healthcheck response[2] time gets higher
+   than normal and the ALB stop serving these instances, effectively making these
+   to stop receiving any request. It affects all instances of the API and Admin
+   which both heavily relies on calls to the same database.
+10. As the database slowed down to answer the API, the latter does not stop
+   receiving and processing notifications. It continues to makes calls to the database
+   and consumes more database connections as previous ones are still busy.
+11. Errors of maximum number of database connections getting reached start
+   to appear at that point.
+12. The API instances are down as these are detached from the load balancers,
+   the database is overloaded with connections and can't take any more. **We are
+   losing notifications** because we don't save these when the API is unavailable.
+
+[1] An individual commit in a database in not a problem usually, but when
+several of these occurs, potentially over 100,000 in the case of Notify in time
+of high traffic, it becomes a performance issue. An apt analogy is possible
+with someone getting to the cashier in a grocery store to pay their items:
+imagine that you would have to pay for each these items individually, i.e.
+a) scan one and only one item, b) ask for the total, c) pay with cash or card,
+d) repeat for all remaining items. This will take quite a long time compared
+to paying for all items in one batch within one transaction. Saving one
+notification every time in the database is akin to paying for one grocery item
+individually in one transaction, except you have potentially thousands of items
+remaining afterward. Ideally, we want to buy/save all for all of these in one
+go.
+
+[2] The EKS environment listens [on the `readinessProbe` health check](https://github.com/cds-snc/notification-manifests/blob/6ac406b315112d2b66d93d8fa74406c3e6e7639f/base/api-deployment.yaml#L216)
+that our setup exposes and it involves a direct database check. When the
+database slows down, so does the probe. The direct reaction from EKS is to
+detach the instances from the application's load balancers.
+
+_TODO: Add sequence or flow chart_
+
+## Options
+
+### Stream the notifications save
+
+Upon receiving a notification from the API, do not save it in the current
+*notification-api* process but defer it to a celery task. This does not
+address the root cause of our issue but offer many benefits, namely:
+
+1. This is an iterative improvement where we send the database save into
+   [Celery land](https://docs.celeryproject.org/en/stable/). Having these into
+   a Celery tasks means we can more easily control the flow of these operations
+   and control these with all of the Celery features. Specifically, we can batch
+   the notifications using [celery-batch](https://pypi.org/project/celery-batches/).
+2. The number of database connections will be lowered in the API component
```

Comment:
I would add such that even if the database is slowed the API route to create notifications will not be slowed causing the EKS chain reaction of detaching the load balancer

Reply:
Good point, although the API could still get detached from the load balancer if the database get slowed down, unless we do the suggestion for *Tweak the readiness and liveness Kubernetes probe*. But the API will be offloaded of DB operations and should process faster.